pkg/oam/util/helper.go

```diff
@@ -633,6 +647,138 @@ func RawExtension2AppConfig(raw runtime.RawExtension) (*v1alpha2.ApplicationConf
 	return ac, nil
 }
 
+// RawExtension2Component converts runtime.RawExtention to Component
+func RawExtension2Component(raw runtime.RawExtension) (*v1alpha2.Component, error) {
+	c := &v1alpha2.Component{}
+	b, err := raw.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(b, c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// ExtractComponentName will extract the componentName from a revisionName
+func ExtractComponentName(revisionName string) string {
+	splits := strings.Split(revisionName, "-")
+	return strings.Join(splits[0:len(splits)-1], "-")
+}
+
+// AppConfig2ComponentManifests convert AppConfig and Components to a slice of ComponentManifest.
+func AppConfig2ComponentManifests(acRaw runtime.RawExtension, comps []common.RawComponent) ([]*oamtypes.ComponentManifest, error) {
+	var err error
+	ac, err := RawExtension2AppConfig(acRaw)
+	if err != nil {
+		return nil, err
+	}
+	cms := make([]*oamtypes.ComponentManifest, len(ac.Spec.Components))
+	for i, acc := range ac.Spec.Components {
+		cm := &oamtypes.ComponentManifest{
+			Name:         acc.ComponentName,
+			RevisionName: acc.RevisionName,
+		}
+		if acc.ComponentName == "" && acc.RevisionName != "" {
+			cm.Name = ExtractComponentName(acc.RevisionName)
+		}
+		for _, compRaw := range comps {
+			comp, err := RawExtension2Component(compRaw.Raw)
+			if err != nil {
+				return nil, err
+			}
+			cm.RevisionHash = comp.GetLabels()[oam.LabelComponentRevisionHash]
+			if comp.Name == acc.ComponentName {
+				cm.Workload, err = RawExtension2Unstructured(&comp.Spec.Workload)
+				if err != nil {
+					return nil, err
+				}
+				if comp.Spec.Helm != nil {
+					rls, _ := RawExtension2Unstructured(&comp.Spec.Helm.Release)
+					if err != nil {
+						return nil, err
+					}
+					repo, _ := RawExtension2Unstructured(&comp.Spec.Helm.Repository)
+					if err != nil {
+						return nil, err
+					}
+					cm.WorkloadAddonResources = []*unstructured.Unstructured{rls, repo}
+				}
+				break
+			}
+		}
+		cm.Traits = make([]*unstructured.Unstructured, len(acc.Traits))
+		for j, t := range acc.Traits {
+			cm.Traits[j], err = RawExtension2Unstructured(&t.Trait)
+			if err != nil {
+				return nil, err
+			}
+		}
+		cm.Scopes = make([]*corev1.ObjectReference, len(acc.Scopes))
+		for x, s := range acc.Scopes {
+			cm.Scopes[x] = &corev1.ObjectReference{
+				Kind:       s.ScopeReference.Kind,
+				Name:       s.ScopeReference.Name,
+				APIVersion: s.ScopeReference.APIVersion,
+			}
+		}
+		cms[i] = cm
+	}
+	return cms, nil
+}
+
+// ComponentManifests2AppConfig convert a slice of ComponentManifest to AppConfig and Components.
+func ComponentManifests2AppConfig(cms []*oamtypes.ComponentManifest) (runtime.RawExtension, []common.RawComponent, error) {
```

Comment:
I didn't see it's being used? And no test cases

Reply:
used in app_controller revision.go. Will add tests for these new utils.